File: Src2/Icm.Core/Basic types extensions/NullableExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Icm

    Public Module NullableExtensions

        ''' <summary>
        ''' Abbreviation of Nullable.Value
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="n"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <Extension()>
        Public Function V(Of T As Structure)(ByVal n As T?) As T
            If n.HasValue Then
                Return n.Value
            Else
                Throw New InvalidOperationException
            End If
            Return n.Value
        End Function

        ''' <summary>
        ''' Abbreviation of "Not var.HasValue"
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="n"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <Extension()>
        Public Function HasNotValue(Of T As Structure)(ByVal n As T?) As Boolean
            Return Not n.HasValue
        End Function


        ''' <summary>
        ''' If a Nullable has value, the value; otherwise, a provided substitution.
        ''' </summary>
        ''' <typeparam name="T"></typeparam>
        ''' <param name="o"></param>
        ''' <param name="subst"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        <Extension()>
        Function IfNull(Of T As Structure)(ByVal o As T?, ByVal subst As Object) As Object
            If o.HasValue Then
                Return o.Value
            Else
                Return subst
            End If
        End Function

    End Module
End Namespace

File: Src/Icm.Core/Basic types extensions/NullableExtensions.cs
using System;

namespace Icm
{
    public static class NullableExtensions
    {
        public static T V<T>(this T? n) where T: struct
        {
            if (n.HasValue)
            {
                return n.Value;
            }

            throw new InvalidOperationException();
        }

        public static bool HasNotValue<T>(this T? n) where T : struct
        {
            return !n.HasValue;
        }


        public static object IfNull<T>(this T? o, object subst) where T : struct
        {
            return o ?? subst;
        }
    }
}